File: Roadmap/07 - PILAS Y COLAS/python/cesar-ch.py
# pila
pila = []

# push
pila.append(1)
pila.append(2)
pila.append(3)

# pop
pila.pop()
print(pila)

# cola
cola = []

# enqueue
cola.append(1)
cola.append(2)
cola.append(3)

# dequeue
cola.pop(0)
print(cola)

# DIFICULTAD EXTRA


def webNavigation():
    stack = []

    while True:
        action = input("Añade una url o interactúa con palabras adelante/atras/salir: ")

        if action == "salir":
            print("Saliendo del navegador web.")
            break
        elif action == "adelante":
            continue
        elif action == "atras":
            if len(stack) > 0:
                stack.pop()
        else:
            stack.append(action)

        if len(stack) > 0:
            print("Has navegado a la web:", stack[-1])
        else:
            print("Estás en la página de inicio.")


webNavigation()


def sharedPrinted():
    queue = []

    while True:
        action = input("Añade un documento o selecciona imprimir/salir: ")

        if action == "salir":
            break
        elif action == "imprimir":
            if len(queue) > 0:
                print("Imprimiendo:", queue.pop(0))
        else:
            queue.append(action)
        print("Cola de impresión:", queue)


sharedPrinted()

File: Roadmap/07 - PILAS Y COLAS/java/cesar-ch.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {

    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        // pila
        List<Integer> pila = new ArrayList<>();

        // push
        pila.add(1);
        pila.add(2);
        pila.add(3);

        // pop
        pila.remove(pila.size() - 1);
        System.out.println(pila);

        // cola
        List<Integer> cola = new ArrayList<>();

        // enqueue
        cola.add(1);
        cola.add(2);
        cola.add(3);

        // dequeue
        cola.remove(0);
        System.out.println(cola);

        // DIFICULTAD EXTRA
        webNavigation();
        sharedPrinted();
    }

    public static void webNavigation() {
        List<String> stack = new ArrayList<>();

        while (true) {
            System.out.print("Añade una url o interactúa con palabras adelante/atras/salir: ");
            String action = scanner.nextLine();

            if (action.equals("salir")) {
                System.out.println("Saliendo del navegador web.");
                break;
            } else if (action.equals("adelante")) {
                continue;
            } else if (action.equals("atras")) {
                if (stack.size() > 0) {
                    stack.remove(stack.size() - 1);
                }
            } else {
                stack.add(action);
            }

            if (!stack.isEmpty()) {
                System.out.println("Has navegado a la web: " + stack.get(stack.size() - 1));
            } else {
                System.out.println("Estás en la página de inicio.");
            }
        }
    }

    public static void sharedPrinted() {
        List<String> queue = new ArrayList<>();

        while (true) {
            System.out.print("Añade un documento o selecciona imprimir/salir: ");
            String action = scanner.nextLine();

            if (action.equals("salir")) {
                break;
            } else if (action.equals("imprimir")) {
                if (!queue.isEmpty()) {
                    System.out.println("Imprimiendo: " + queue.remove(0));
                }
            } else {
                queue.add(action);
            }
            System.out.println("Cola de impresión: " + queue);
        }
    }
}
